File: howl-application/src/main/java/io/quartic/howl/StorageBackendConfig.java
package io.quartic.howl;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import io.quartic.howl.storage.StorageBackend;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property="type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = GcsStorageBackendConfig.class, name = "google_cloud_storage"),
        @JsonSubTypes.Type(value = DiskStorageBackendConfig.class, name = "local_disk"),
})
public interface StorageBackendConfig {
    StorageBackend build();
}

File: howl-application/src/main/kotlin/io/quartic/howl/StorageBackendConfig.kt
package io.quartic.howl

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonSubTypes.Type
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeInfo.As
import com.fasterxml.jackson.annotation.JsonTypeInfo.Id
import io.quartic.howl.storage.StorageBackend

@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type")
@JsonSubTypes(
        Type(value = GcsStorageBackendConfig::class, name = "google_cloud_storage"),
        Type(value = DiskStorageBackendConfig::class, name = "local_disk")
)
interface StorageBackendConfig {
    fun build(): StorageBackend
}
